tensorboard/components/tensor_widget/types.ts

```diff
@@ -0,0 +1,196 @@
+    number|number[]|number[][]|number[][][]|
+    string|string[]|string[][]|string[][][];
+
+/**
+ * A data structure that keeps track of how an n-dimensional array (tensor)
+ * is sliced down to a smaller number of dimensions for visualization
+ * in the tensor widget.
+ *
+ * For example, suppose there is a 4D tensor of shape [16, 128, 128, 3]
+ * representing a NHWC image batch. If you'd like to get the 4-by-3 top-left
+ * corner of the first image of the last color channel  displayed in the tensor widgete, this interface
+ * should have the following concrete value:
+ *
+ * ```
+ * {
+ *   slicingDimsAndIndices: [{
+ *     dim: 0,
+ *     index: 0
+ *   }, {
+ *     dim: 3,
+ *     index: 2
+ *   }],
+ *   viewingDims: [1, 2],
+ *   verticalRange: [0, 4],
+ *   horizontalRange: [0, 3]
+ * }
+ * ```
+ */
+export interface TensorViewNavigationStatus {
+  /**
+   * Which dimensions of the tensor are sliced down to a slice of 1.
+   *
+   * - The `dim` field is the 0-based dimension index.
+   * - The `index` is the 0-based index for the selected slice.
+   */
+  slicingDimsAndIndices: Array<{dim: number, index: number}>;
+
+  /**
+   * Which dimensoins are used for viewing (i.e., rendered in the
+   * tensor widget, as a table, by default.)
+   */
+  viewingDims: number[];
+
+  /**
+   * The indices from the first viewing dimension, which are shown as rows.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  verticalRange: [number, number];
+
+  /**
+   * The indices from the second viewing dimension, which are shown as columns.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  horizontalRange: [number, number];
+
+  /**
+   * Optional dimension for depth.
+   *
+   * This supports visualization that requires a depth dimension, e.g.,
+   * color channels.
+   */
+  depthDim?: number;
+}
+
+/** Options used during the creation of a single-tensor tensor widget. */
+export interface SingleTensorWidgetOptions {
+  /**
+   * Name of the tenosr.
+   *
+   * Optional. If provided, it will be shown on the widget.
+   * If not provided, no name will be shown.
+   */
+  name?: string;
+
+  /**
+   * Whether the health-pill portion of the tensor widget is to be
+   * included
+   *
+   * Defaults to `true`.
+   */
+  includeHealthPill?: boolean;
+
+  /** Defaults to `true`. */
+  includeMenu?: boolean;
+
+  /**
+   * How many decimal places to display the values in.
+   *
+   * The values of the tensor may be displaced in the decimal notation, or
+   * the engineering notation, depending automatically by the tensor-widget
+   * library based on the maximum absolute value of the elements of the tensor.
+   */
+  decimalPlaces?: number;
+
+  /** TODO(cais): Add support for custom tensor renderers. */
+}
+
+/**
+ * A TensorWidget that interactively visuzlies a single tensor.
+ */
+export interface SingleTensorWidget {
```

Comment:
I wonder whether maybe it's simpler to just call this TensorWidget, and at the point at which a multi-tensor widget is needed just to call it e.g. MultipleTensorWidget or DoubleTensorWidget or whatever?  It seems like the singleton case will be far more common so it feels reasonable to make that the implicit assumption when just talking about a TensorWidget.

Ditto for SingleTensorWidgetOptions.

Reply:
Good point. I like the idea. Done. We do plan to support two-tensor diffing views. But those can be called TwoTensorDiffWidget or something like that later.